### ISSUE TITLE
data does not contain any valid RSA or ECDSA certificates

### ISSUE THREAD

**Issue by mmack** [kubevirt/kubevirt#2049]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
Fresh install of kubevirt on a coreos k8s cluster. Version 13 and 14.

**What you expected to happen**:
virt-api pods are ready

**How to reproduce it (as minimally and precisely as possible)**:
I don't know if this is related to coreos, seems like...

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
`tested it with 14 and 13`
- Kubernetes version (use `kubectl version`):
``` Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}    
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-10T23:28:14Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}
``` 
- VM or VMI specifications:
`coreos nodes running on hardware and vmware combined`
- OS (e.g. from /etc/os-release):
```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1967.4.0
VERSION_ID=1967.4.0
BUILD_ID=2019-01-28-0558
PRETTY_NAME="Container Linux by CoreOS 1967.4.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
``` 
- Kernel (e.g. `uname -a`):
`Linux k8s-worker-24.lxdev.domain.com 4.14.96-coreos #1 SMP Mon Jan 28 05:16:10 -00 2019 x86_64 Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz GenuineIntel GNU/Linux
`
- Openssl version
`OpenSSL 1.0.2p  14 Aug 2018`

I spinned up the pod with another entrypoint... The ssl certificate virt-api complains is broken:

```
[virt-api@virt-api-854bd67fd5-s9h5w ~]$ /usr/bin/virt-api 
{"component":"virt-api","level":"warning","msg":"Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.","pos":"client_config.go:548","timestamp":"2019-02-21T12:59:33.788776Z"}
{"component":"virt-api","level":"info","msg":"kubevirt-config ConfigMap does not exist. Using defaults.","pos":"config-map.go:75","timestamp":"2019-02-21T12:59:33.810683Z"}
{"component":"virt-api","level":"warning","msg":"Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.","pos":"client_config.go:548","timestamp":"2019-02-21T12:59:33.810780Z"}
{"component":"virt-api","level":"warning","msg":"Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.","pos":"client_config.go:548","timestamp":"2019-02-21T12:59:33.811464Z"}
panic: error reading /tmp/certsdir847514059/clientCA.crt: data does not contain any valid RSA or ECDSA certificates

goroutine 1 [running]:
kubevirt.io/kubevirt/pkg/virt-api.(*virtAPIApp).Run(0xc000118000)
	/root/go/src/kubevirt.io/kubevirt/pkg/virt-api/api.go:1031 +0x31a
kubevirt.io/kubevirt/pkg/virt-api.(*virtAPIApp).Execute(0xc000118000)
	/root/go/src/kubevirt.io/kubevirt/pkg/virt-api/api.go:170 +0x1c3
main.main()
	/root/go/src/kubevirt.io/kubevirt/cmd/virt-api/virt-api.go:36 +0xc8
```

```
[virt-api@virt-api-854bd67fd5-s9h5w ~]$ openssl x509 -in /tmp/certsdir847514059/clientCA.crt -text -noout
unable to load certificate
140510704424768:error:0906D066:PEM routines:PEM_read_bio:bad end line:crypto/pem/pem_lib.c:794:
``` 

```
[virt-api@virt-api-854bd67fd5-s9h5w ~]$ cat /tmp/certsdir847514059/clientCA.crt
-----BEGIN CERTIFICATE-----
MIIDdjCCAl6gAwIBAgIUFmcsO4kU1gOuC5HiW0SP7y6w3dUwDQYJKoZIhvcNAQEL
...
rhNDY6GLdsONY1rck/hn6etVlGbRFobGz8Q=
-----END CERTIFICATE----------BEGIN CERTIFICATE-----
MIIDdjCCAl6gAwIBAgIUFmcsO4kU1gOuC5HiW0SP7y6w3dUwDQYJKoZIhvcNAQEL
...
rhNDY6GLdsONY1rck/hn6etVlGbRFobGz8Q=
-----END CERTIFICATE-----```

It's the same cert twice.

**Comment by rmohr** [comment]
@mmack could you run `kubectl delete secrets kubevirt-virt-api-certs` and delete all virt-api pods? They should re-create it and start again. It looks like a broken certificate landed in that secret. I wonder how.

**Comment by dcritch** [comment]
I'm getting the same error. I deleted the secret and deleted/recreated the pods but the error persists. Using openshift-intaller+libvirt

**Comment by rmohr** [comment]
@davidvossel do you have any insights?

**Comment by davidvossel** [comment]
whoa, this is strange. There's nothing about cert generation that should be related to the kubernetes version. This is all self contained within KubeVirt's virt-api component. I wouldn't expect it to be possible to get all the way to starting the TLS server if the certs weren't generated correctly. looking at the code this seems impossible unless the k8s secret was mutated somehow. 

assuming this is a dev cluster that you aren't worried about credentials for, I'd be curious to see the ```kubectl get secrets -n kubevirt kubevirt-virt-api-certs -o yaml``` output that is generated.

**Comment by rmohr** [comment]
@mmack @dcritch  so far we never managed to reproduce this. Is this still an issue?

**Comment by dcritch** [comment]
@rmohr I have not seen it reoccur myself. To me, I'd say its no longer an issue

**Comment by terinjokes** [comment]
This looks to have been fixed in 0.17.0. In 0.16.0 the `client-ca-file` datum of the "extension-apiserver-authentication" ConfigMap was prepended with the `requestheader-client-ca-file` datum.

Except there was no newline between the two certificates. This is shown in the example above:

```
-----END CERTIFICATE----------BEGIN CERTIFICATE-----
```

You can see this behavior in [this repl].(https://repl.it/@TerinStock/ThinEarlyMicrostation)

AFAICT, this has nothing to do with the self-signed certificates virt-api created for webhooks and apiservices.

**Comment by rmohr** [comment]
@terinjokes thanks.

/close
